Java06. ДЗ 02, Мордберг Антон

AntonGitName opened the pull request:
Лучше поздно, чем никогда. (с)
@jzuken

AntonGitName committed:
Update gitingore

AntonGitName committed:
add gradle

AntonGitName committed:
JCommander added

AntonGitName committed:
guava chosen for hashing + added last commands declaration

AntonGitName committed:
stage phase added

AntonGitName committed:
init + add

AntonGitName committed:
added status

AntonGitName committed:
added commit

AntonGitName committed:
added log

AntonGitName committed:
added clean

AntonGitName committed:
added reset

AntonGitName committed:
added remove

AntonGitName committed:
moved gradle to top level dir

AntonGitName committed:
duplicate revisions fix

AntonGitName committed:
minor changes

AntonGitName committed:
branches add

AntonGitName committed:
fukk restructure

AntonGitName committed:
tests added + init moved to repository

AntonGitName committed:
all tests + merge

AntonGitName committed:
unused code removed

AntonGitName committed:
added gradle wrapper

AntonGitName committed:
minor changes

AntonGitName committed:
minor changes in cli

AntonGitName committed:
invisible internals

AntonGitName committed:
minor changes in gradle files

AntonGitName committed:
minor changes in gradle files

AntonGitName committed:
Merge branch 'master' into vcs

# Conflicts:
#	build.gradle
#	gradle/wrapper/gradle-wrapper.properties
#	settings.gradle

AntonGitName committed:
minor changes in gradle

jzuken commented on lab-vcs/src/main/java/ru/mit/spbau/antonpp/vcs/core/log/CommitInfo.java:
```diff
@@ -0,0 +1,74 @@
+
+    private String author;
+    private String time;
+    @Nullable
```
это поле nullable, а остальные что?

AntonGitName replied:
Я использовал Nullable/NotNull для того, чтобы самому не забыть, что какие-то поля могу иметь особый смысл (например, что-то может быть null из тех соображений, что это долго считать каждый раз и поэтому их подсчёт можно закэшировать в геттере (который уже возвращает NotNull), или как тут, глобально мне всё равно какое сообщение запишется в лог, но часто есть возможность сгенерировать этот текст, поэтому я так отметил это поле, чтобы не забыть написать/вызвать метод для генерации дефолтного сообщения). По хорошему, все переменные имеет смысл проанотировать, но часто это излишне и не приносит никакой полезной информации (и когда весь код в аннотациях, то это отвлекает и не даёт сконцентрироваться на тех местах, где это действительно важно), а не редко это ещё и ведёт к бессмысленным варнингам в идее (например, у меня есть класс, который создается только через deserialize и поэтому у него пустой конструктор, и если в нем пометить поле как NotNull, то идея всегда будет ругаться на то, что конструктор это поле не инициализировал).

jzuken replied:
я согласен почти со всем тут. просто одна аннотация среди кучи полей выглядит странновато :)

jzuken commented on lab-vcs/src/main/java/ru/mit/spbau/antonpp/vcs/core/log/CommitInfo.java:
```diff
@@ -0,0 +1,74 @@
+import java.io.Serializable;
+
+/**
+ * @author Anton Mordberg
```
хорошо бы более адекватные комментарии всё же

AntonGitName replied:
Ох, на самом деле не хотелось писать джавадок для всего кода - я старался делать структуру прозрачно понятной, но, возможно, в некоторых местах нужно написать какие контракты соблюдаются. Добавлю джавадок к core классам.

jzuken commented:
раз мы делаем консольную утилиту, хорошо бы в конфиг добавить что-то про сборку jar-файла

AntonGitName commented on lab-vcs/src/main/java/ru/mit/spbau/antonpp/vcs/core/Repository.java:
```diff
@@ -0,0 +1,360 @@
+            try {
+                revName = branchResolver.getBranchHead(revName);
+            } catch (BranchException e) {
+                // impossible
```
Да, а что тут такого? getBranchHead кидает исключение только в том случае, если ветки нет, а я только что проверил что она есть, поэтому исключение не может быть никогда (можно, конечно, придумать изотерический случай с многопоточностью или refletion, но тут нет же такого). Хорошо, уберу if перенеся else блок в catch.

jzuken replied:
да, лучше уж так :)

AntonGitName commented:
@jzuken 
> раз мы делаем консольную утилиту, хорошо бы в конфиг добавить что-то про сборку jar-файла

Не очень понял, о чём речь. Хочется в градле сделать таску deploy? Или про какой конфиг речь? Вообще, сейчас предполагается такая сборка: `./gradlew :lab-vcs:jar`

AntonGitName committed:
repository and branchResolver refactored

 Repository code simplified with few shortcut methods. All checks from BranchResolver removed and now calling code must make sure that branch exists(does not exist) before deleting (adding) a branch.

AntonGitName committed:
repository and branchResolver refactored + lombok added + Revision classes restructured and optimized

 Repository code simplified with few shortcut methods. All checks from BranchResolver removed and now calling code must make sure that branch exists(does not exist) before deleting (adding) a branch.

jzuken commented:
я понимаю, но манифест оно туда не кладёт: no main manifest attribute, in AuJava2-1.0.jar

AntonGitName committed:
Repository javadoc

AntonGitName committed:
FileSerializable javadoc

AntonGitName committed:
Status javadoc

AntonGitName committed:
RevisionDiff javadoc

AntonGitName committed:
FileStatus javadoc

AntonGitName committed:
WorkingDir javadoc

AntonGitName committed:
Stage javadoc

AntonGitName committed:
Revision javadoc

AntonGitName committed:
Commit javadoc

AntonGitName committed:
revisions javadoc

AntonGitName committed:
log package javadoc

AntonGitName committed:
branchResolver javadoc

AntonGitName committed:
useless javadoc

AntonGitName committed:
updated gradle -> now deploy available

AntonGitName committed:
updated gradle

AntonGitName committed:
changes log4j log file name

AntonGitName commented:
@jzuken 
Больше похоже на то, что Вы написали `./gradlew jar` и собрали не тот модуль. Там всё в порядке с манифестом.
Короче, я добавил таску deploy и в корневой проект, и в модуль. Так что `./gradlew deploy` точно всё соберёт и напишет где jar лежит.
Добавил ещё кучу джавадоков и немножко код отшлифовал.

jzuken commented:
не, я всё запускал, как написано :)
ок, гляну. вы только сделайте сабмиты с новыми версиями на hwproj, чтобы я не пропускал решения

jzuken commented on lab-vcs/src/test/java/ru/mit/spbau/antonpp/vcs/core/RepositoryTest.java:
```diff
@@ -0,0 +1,510 @@
+import ru.mit.spbau.antonpp.vcs.core.exceptions.CheckoutException;
+import ru.mit.spbau.antonpp.vcs.core.exceptions.ResetException;
+import ru.mit.spbau.antonpp.vcs.core.exceptions.SerializationException;
+import ru.mit.spbau.antonpp.vcs.core.log.LogRecord;
```
а куда этот файл делся?

AntonGitName replied:
Он попал в gitignore из-за того, что лежал в пакете log  ><

AntonGitName committed:
thx gitignore

AntonGitName committed:
minor changes

jzuken commented:
Created jar in folder `/home/me/git/au/mordberg/lab-vcs/target`
:lab-vcs:createTargetDir UP-TO-DATE
:lab-vcs:compileJava UP-TO-DATE
:lab-vcs:processResources UP-TO-DATE
:lab-vcs:classes UP-TO-DATE
:lab-vcs:jar UP-TO-DATE
:lab-vcs:deploy UP-TO-DATE
:deploy UP-TO-DATE

BUILD SUCCESSFUL

Total time: 0.939 secs
➜  me@faraway ~/git/au/mordberg git:(vcs) ✗ java -jar build/libs/AuJava2-1.0.jar                                                                               18:24:28
no main manifest attribute, in build/libs/AuJava2-1.0.jar


что я делаю не так? :)

jzuken commented:
а, я не оттуда jar запускал

jzuken commented:
log/vcs.log, я думаю, не стоит показывать в untracked, раз он служебный

jzuken commented:
а как переключиться на новый бранч? я сделал новый new, пишу ✗ java -jar lab-vcs-1.0.jar checkout new, оно мне говорит Could not find single commit. Found: []

jzuken commented:
и круто было бы где-то показывать текущий бранч. например при вызове branch без параметров

AntonGitName commented:
@jzuken 
> и круто было бы где-то показывать текущий бранч. например при вызове branch без параметров
status показывает текущую ветку.

> а как переключиться на новый бранч? я сделал новый new, пишу ✗ java -jar lab-vcs-1.0.jar checkout new, оно мне говорит Could not find single commit. Found: []
branch сразу делает и checkout (и не даст сделать, если есть не закоммиченные файлы). Поэтому checkout и не находил указанную ветку. Вообще, новая ветка это отдельная тема - она ещё не успела попасть в BranchResolver и единственное места, где про неё есть запись это текущий Stage. А все ветки, в которых есть коммиты, они уже в резолвере записаны.

jzuken commented:
ага, понял, спасибо

jzuken commented:
✗ java -jar lab-vcs-1.0.jar merge new                                                                 18:53:59
Failed to merge.

очень информативно. и главное, сразу понятно, что дальше делать :)

jzuken commented:
если что, я добавил один файл в мастер, потом сделал новый бранч, добавил другой файл туда, вернулся в мастер и попытался замерджить их

jzuken commented:
кусок лога (как хорошо, что он есть):

2016-11-05 18:54:06 ERROR AbstractCommand:80 - Failed to merge.
ru.mit.spbau.antonpp.vcs.core.exceptions.MergeException: Failed to commit merge
        at ru.mit.spbau.antonpp.vcs.core.revision.Stage.merge(Stage.java:310)
        at ru.mit.spbau.antonpp.vcs.core.Repository.merge(Repository.java:225)
        at ru.mit.spbau.antonpp.vcs.cli.commands.CommandMerge.runInternal(CommandMerge.java:41)
        at ru.mit.spbau.antonpp.vcs.cli.commands.AbstractCommand.run(AbstractCommand.java:37)
        at ru.mit.spbau.antonpp.vcs.Application.run(Application.java:92)
        at ru.mit.spbau.antonpp.vcs.Application.<init>(Application.java:62)
        at ru.mit.spbau.antonpp.vcs.Application.main(Application.java:67)
Caused by: ru.mit.spbau.antonpp.vcs.core.exceptions.CommitException: Failed to move file
        at ru.mit.spbau.antonpp.vcs.core.revision.Stage.commit(Stage.java:148)
        at ru.mit.spbau.antonpp.vcs.core.revision.Stage.merge(Stage.java:308)
        ... 6 more
Caused by: java.nio.file.FileAlreadyExistsException: /home/me/git/au/mordberg/lab-vcs/target/.vcs-internals/revs/f591dec073b1e471e416ee8cb979854b/files/2d94eda4319213bba047ca5fa8ce2d46
        at sun.nio.fs.UnixCopyFile.copy(UnixCopyFile.java:551)

AntonGitName committed:
merge & commit bugfixes

AntonGitName committed:
branch name bugfixes + log & jar ignored by vcs

jzuken commented:
положил в мастер файл, создал второй бранч, положил файл туда вторым коммитом, возвращаюсь чекаутом в мастер, а там лог оба коммита показывает, и оба файла лежат на диске. так задумано?

AntonGitName committed:
merged-in master

AntonGitName committed:
merged-in master

AntonGitName committed:
checkout now cleans repo + log filters records

AntonGitName commented:
@jzuken

У чекаута и планировалось такое поведение (файлы, которые не в ревизии, никак не трогаются и не удаляются при чекауте) Я изменил, чтобы после чекаута сразу clean вызывался и ревизия чистенькая была. Логи тоже исправил - теперь фильтрую коммиты, которые никак не относятся к текущей ревизии.

jzuken commented:
ок, покатит
